app/src/main/java/com/example/tasky/feature_authentication/data/repository/AuthRepositoryImpl.kt

```diff
@@ -0,0 +1,68 @@
+package com.example.tasky.feature_authentication.data.repository
+
+import android.content.SharedPreferences
+import com.example.tasky.feature_authentication.data.remote.TaskyAuthApi
+import com.example.tasky.feature_authentication.data.remote.auth.LoginRequest
+import com.example.tasky.feature_authentication.data.remote.auth.RegistrationRequest
+import com.example.tasky.feature_authentication.data.util.AuthResult
+import com.example.tasky.feature_authentication.domain.repository.AuthRepository
+import retrofit2.HttpException
+
+class AuthRepositoryImpl(
+    private val api: TaskyAuthApi,
+    private val prefs: SharedPreferences
```

Comment:
I'm a fan of also abstracting the preferences, since that's something you can implement in various ways. Having an abstraction allows you to easily swap it out (for example if you decide to use DataStore in future instead of SharedPref)

The following allows you to easily access user data from any layer in your app while being able to also easily swap out preferences implementations. The only thing needed for that would be to swap out the shared preferences object in the `UserPreferencesImpl` constructor with let's say a DataStore instance.

*Domain model*
```
data class AuthenticatedUser(
    val fullName: String,
    val email: String,
    val token: String
)
```

*Domain abstraction*
```
interface UserPreferences {
    fun saveAuthenticatedUser(user: AuthenticatedUser)
    fun getAuthenticatedUser(): AuthenticatedUser?
    fun clearPreferences()
}
```

*Data implementation*
```
class UserPreferencesImpl(
    private val sharedPref: SharedPreferences
): UserPreferences {
    override fun saveAuthenticatedUser(user: AuthenticatedUser) {
        // Save JSON of user in shared pref
    }

    override fun getAuthenticatedUser(): AuthenticatedUser? { ... }

    override fun clearPreferences() { ... }
}
```

In your Dagger module:
```
@Provides
@Singleton
fun provideUserPreferences(sharedPreferences: SharedPreferences): UserPreferences {
    return UserPreferencesImpl(sharedPreferences)
}
```

Then, you can inject this wherever you want.